File: src/item.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone, Copy, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum ItemKind {
    /// Article with url and content
    #[default]
    Article,
    /// Bookmark with url only
    Bookmark,
    /// Note with markdown content
    Note,
    /// Task with markdown content
    Task,
    /// Video with youtube url and transcription
    Video,
}

#[derive(Debug, Serialize, Deserialize, Clone, Copy, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum ItemStatus {
    /// Default for all items
    #[default]
    New,
    /// Planned to read, watch or to do
    Planned,
    /// In progress or reading, watching
    InProgress,
    Paused,
    /// Item is done
    Done,
    /// Item is not relevant anymore
    Rejected,
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct ItemMetadata {
    pub kind: ItemKind,
    pub title: String,
    pub url: Option<String>,
    pub tags: Vec<String>,
    pub author: Option<String>,
    pub status: ItemStatus,
    pub created_at: String,
    pub updated_at: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, Default)]
pub struct Item {
    pub kind: ItemKind,
    pub path: String,
    pub title: String,
    pub url: Option<String>,
    pub author: Option<String>,
    pub tags: Vec<String>,
    pub status: ItemStatus,
    pub created_at: String,
    pub updated_at: Option<String>,
}

impl From<Item> for ItemMetadata {
    fn from(item: Item) -> Self {
        Self {
            kind: item.kind,
            title: item.title,
            url: item.url,
            tags: item.tags,
            author: item.author,
            status: item.status,
            created_at: item.created_at,
            updated_at: item.updated_at,
        }
    }
}

impl Item {
    pub fn from_metadata(m: ItemMetadata, path: String) -> Self {
        Self {
            kind: m.kind,
            path,
            title: m.title,
            url: m.url,
            tags: m.tags,
            author: m.author,
            status: m.status,
            created_at: m.created_at,
            updated_at: m.updated_at,
        }
    }
}

File: frontend/src/api/Item.ts
export type ItemKind = "article" | "note" | "task" | "video";
export type ItemStatus =
  | "new"
  | "planned_high"
  | "planned"
  | "planned_low"
  | "in_progress"
  | "paused"
  | "done"
  | "rejected";

export interface Item {
  path: string;
  assetPath?: string;
  url?: string;
  title?: string;
  author?: string;
  name: string;
  assets: string[];
  attachmentsMetadata: Record<string, AttachmentMetadata>;
  tags: string[];
  created_at: string;
  kind?: ItemKind;
  status?: ItemStatus;
}

export type ItemResp =
  | { ok: true; items: Item[]; total: number }
  | { ok: false; error: string };

export type CreateResp =
  | { ok: true; item?: Item }
  | { ok: false; error: string };

export interface CreateItemPayload {
  kind: ItemKind;
  title?: string;
  url?: string;
  tags?: string;
  content?: string;
  created_at?: string;
}

export type AssetResp =
  | { ok: true; body: string; contentType: string }
  | { ok: false; status?: number; error: string };

export interface AttachmentMetadata {
  name: string;
  size: number;
}

export type Attachment = {
  path: string;
  content: boolean;
  note: boolean;
  attachments: string[];
  metadata: AttachmentMetadata[];
};

export type AttachmentsResp =
  | { ok: true; path: string; attachments: string[]; metadata: AttachmentMetadata[] }
  | { ok: false; error: string };

export type UploadResp = { ok: true } | { ok: false; error: string };

export type UpdateTagsResp = { ok: true } | { ok: false; error: string };

/** All known tags with their usage counts, from `GET /api/tags`. */
export type TagsResp =
  | { ok: true; tags: Record<string, number> }
  | { ok: false; error: string };

export type TagFreq = { name: string; count: number };

export type UpdateStatusResp = { ok: true } | { ok: false; error: string };

export interface UpdateItemPayload {
  title: string;
  created_at: string;
  author?: string;
}

export type UpdateItemResp =
  | { ok: true; item?: Item }
  | { ok: false; error: string };

export type DeleteItemResp = { ok: true } | { ok: false; error: string };

export type DeleteAttachmentResp = { ok: true } | { ok: false; error: string };

/** AI-generatable content features, driving `PUT /api/items/{item_path}/ai`. */
export type AiFeature = "summary" | "mindmap";

export type GenerateAiResp = { ok: true } | { ok: false; error: string };

/** Query for tag-based filtering: included tags are AND-ed, excluded tags are `!`-prefixed. */
export interface ItemQuery {
  tags?: string[];
  exclude?: string[];
  kind?: ItemKind;
  status?: ItemStatus;
  /** Single-day filter in `YYYY-MM-DD` format. */
  date?: string;
  /** Max number of items to return. */
  limit?: number;
  /** Number of items to skip, for pagination. */
  offset?: number;
}
